jwst/stpipe/utilities.py

```diff
@@ -144,3 +148,61 @@ def folder_traverse(folder_path, basename_regex='.+', path_exclude_regex='^$'):
         for file in files:
             if basename_regex.match(file):
                 yield os.path.join(root, file)
+
+
+def record_step_status(datamodel, cal_step, success=True):
+    """Record whether or not a step completed in meta.cal_step
+
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to modify in place
+
+    cal_step : str
+        The attribute in meta.cal_step for recording the status of the step
+
+    success : bool
+        If True, then 'COMPLETE' is recorded.  If False, then 'SKIPPED'
+    """
+    if success:
+        status = COMPLETE
+    else:
+        status = SKIPPED
+
+    if isinstance(datamodel, Sequence):
+        for model in datamodel:
+            model.meta.cal_step._instance[cal_step] = status
+    else:
+        datamodel.meta.cal_step._instance[cal_step] = status
+
+    # TODO: standardize cal_step naming to point to the official step name
+
+
+def query_step_status(datamodel, cal_step):
+    """Query the status of a step in meta.cal_step
+    
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` or `~jwst.datamodels.ModelContainer` instance
+        The datamodel or container of datamodels to check
+        
+    cal_step : str
+        The attribute in meta.cal_step to check
+    
+    Returns
+    -------
+    status : str
+        The status of the step in meta.cal_step, typically 'COMPLETE' or 'SKIPPED'
+
+    Notes
+    -----
+    In principle, a step could set the COMPLETE status for only some subset 
+    of models, so checking the zeroth model instance may not always be correct.
+    However, this is not currently done in the pipeline. This function should be
+    updated to accommodate that use-case as needed.
+    """
+    if isinstance(datamodel, Sequence):
+        statuses = [getattr(model.meta.cal_step, cal_step, NOT_SET) for model in datamodel]
+        return statuses[0]
```

Comment:
```suggestion
        return query_step_status(model[0], cal_step)
```
This avoids iterating through all of the models.

Reply:
Any objection to this change?

Reply:
no objections